crates/apollo-router-core/src/query_planner/model.rs

```diff
@@ -70,6 +75,185 @@ impl PlanNode {
             Self::Flatten(flatten) => Box::new(flatten.node.service_usage()),
         }
     }
+
+    /// Recursively validate a query plan node making sure that all services are known before we go
+    /// for execution.
+    ///
+    /// This simplifies processing later as we can always guarantee that services are configured for
+    /// the plan.
+    ///
+    /// # Arguments
+    ///
+    ///  *   `plan`: The root query plan node to validate.
+    pub fn validate_services_against_plan(
```

Comment:
do we really need to do that? Since the query plan was generated from the schema, and the service registry was generated from the schema too, we're sure they will all be there. Or are there other cases that we should prepare for? Like circuit breaking for a failing service?

Reply:
Again this is old code. You're probably right that when the query will be parsed and validated against schema we won't need this anymore.

Reply:
(Do you mind creating an issue from this comment? :grin: )

Reply:
the [bridge planner](https://github.com/apollographql/federation/blob/main/router-bridge/js-src/plan.ts#L18) calls functions that validate the [operation against the schema](https://github.com/apollographql/federation/blob/3b2f486e876f8da8a66e3e94bfae8d8db0c2d2cc/internals-js/src/operations.ts#L1257) as well as the fragments within, before building the query plan.

It might be worth a test though.

Reply:
https://github.com/apollographql/router/issues/177